language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionNodeAnalyzer.java

```diff
@@ -0,0 +1,259 @@
+import io.ballerina.compiler.syntax.tree.EnumDeclarationNode;
+import io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
+import io.ballerina.compiler.syntax.tree.MethodDeclarationNode;
+import io.ballerina.compiler.syntax.tree.ModulePartNode;
+import io.ballerina.compiler.syntax.tree.ModuleVariableDeclarationNode;
+import io.ballerina.compiler.syntax.tree.ModuleXMLNamespaceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NodeList;
+import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.Token;
+import io.ballerina.compiler.syntax.tree.TypeDefinitionNode;
+import org.ballerinalang.langserver.codeaction.NodeBasedPositionDetailsImpl.PositionDetailsBuilder;
+
+import java.util.Optional;
+
+/**
+ * Node analyzer for the code actions.
+ * This will calculate the node information and fill the {@link PositionDetailsBuilder} and the information will be
+ * reflected to the code actions via the PositionDetails API.
+ *
+ * @since 2201.1.x
+ */
+public class CodeActionNodeAnalyzer extends NodeVisitor {
+    private final PositionDetailsBuilder positionDetailsBuilder;
+    private final int positionOffset;
+
+    public CodeActionNodeAnalyzer(PositionDetailsBuilder positionDetailsBuilder, int positionOffset) {
+        this.positionDetailsBuilder = positionDetailsBuilder;
+        this.positionOffset = positionOffset;
+    }
+
+    @Override
+    public void visit(ServiceDeclarationNode node) {
+        int serviceKwStart = node.serviceKeyword().textRange().startOffset();
+        int openBraceEnd = node.openBraceToken().textRange().endOffset();
+        int closeBraceEnd = node.closeBraceToken().textRange().endOffset();
+
+        if (isWithinRange(serviceKwStart, closeBraceEnd)) {
+            positionDetailsBuilder.setEnclosingDocumentableNode(node);
+            if (isWithinRange(serviceKwStart, openBraceEnd)) {
+                positionDetailsBuilder.setDocumentableNode(node);
+                return;
+            }
+        }
+
+        this.visitNode(node);
+    }
+
+    @Override
+    public void visit(FunctionDefinitionNode node) {
+        if (node.functionBody().kind() != SyntaxKind.FUNCTION_BODY_BLOCK) {
+            return;
+        }
+
+        NodeList<Token> qualifiers = node.qualifierList();
+        int startOffset = !qualifiers.isEmpty() ?
+                qualifiers.get(0).textRange().startOffset() :
+                node.functionKeyword().textRange().startOffset();
+        int openBraceEnd = ((FunctionBodyBlockNode) node.functionBody()).openBraceToken().textRange().endOffset();
+        int closeBraceEnd = ((FunctionBodyBlockNode) node.functionBody()).closeBraceToken().textRange().endOffset();
```

Comment:
We can get `(FunctionBodyBlockNode) node.functionBody()` to a variable and use it so this calculation is not repeated twice

Reply:
Done